### PR TITLE
feat: List of addresses from the wallet

### PR DESCRIPTION
**Pull request by pedroferreira1**
- [x] Remove debug logs;
- [x] Add comments in methods/states/propTypes;
- [x] Add link to explorer address search screen on address row click.

Unresolved questions:

1. Should we have a new item on the menu? At first I tried to add some link/button close to the address box in the main wallet screen but none of my attempts were good, so I just added this new menu.
2. I created a new pagination because we already have all the elements on the list and I always think it's better to have all the pages there (if it's possible). Should we keep the same pagination as in the other screens with 'Previous'/'Next'?
3. Should we show all addresses including the ones from the `GAP_LIMIT`. I think we should because maybe the user has generated new addresses, sent to someone but still haven't received the txs, in that case none of those addresses would appear on the list. However, showing all addresses might be confusing to the user.

image
image

### CODE REVIEW

**Comment by msbrogli** [comment]
Instead of used, I'd include four columns: Number of Transactions, Total Received, Total Sent, and Balance. If it is too much, we can keep just Number of Transactions and Balance. What do you think?

**Comment by msbrogli** [comment]
I'm not sure whether it should be a menu. Maybe we can put a "Show all addresses" next to the "Address to receive tokens" box.

**Comment by msbrogli** [comment]
About the GAP limit, we can include an explanation that the last 20 addresses must be empty. We could also add a separator like "------- (show the last 20 addresses) -------", toggling to hide when clicked.

**Comment by obiyankenobi** [comment]
I wonder if it's useful for users to know the index. I think it might be more confusing. 

Regarding where to access this page, I don't think we should place it on the menu or the main page. This looks like something people won't use often. I'd rather have just a link on the settings page, even though it's not a setting option.

**Comment by msbrogli** [comment]
@obiyankenobi The index is useful to generate the address derivation path. Maybe the user might have a way to see the full path.

**Comment by pedroferreira1** [comment]
> Instead of used, I'd include four columns: Number of Transactions, Total Received, Total Sent, and Balance. If it is too much, we can keep just Number of Transactions and Balance. What do you think?

@msbrogli I don't like it for the same reason we made the explorer 'Address Search' page so complex. An address might have many tokens and the balance is different for each other. The best option would be to have an address screen similar to the one from the explorer, but this is lot of work for now. I've changed the 'used' column for the 'number of transactions' column and added a link to the explorer, so the user can see all the details of the address. What do you think? 

> I don't think we should place it on the menu or the main page. This looks like something people won't use often. I'd rather have just a link on the settings page, even though it's not a setting option.

I don't know about the Settings page. I don't like it on the menu either and I think the best place would be close to the address box. I will try some options and paste here so we can discuss. If we don't find a good fit, I think the Settings is a better place than the menu.

> What about a toggle do show/hide unused addresses? We could also have a "?" icon next to it with more information.

The problem is not the unused addresses, the problem to me is the addresses from the gap limit. That most users won't use but some will, so it's gonna be weird for most of them to see 20 unused addresses in the end and some will be okay because they will understand. Maybe we should just add this '?' explaining about the last 20 unused addresses.

**Comment by pedroferreira1** [comment]
Options of places to add a button/link to go to Address List screen.

1. 
image

2. 
image

3. 
image

4. 
image

5. 
image

What do you think? @msbrogli @obiyankenobi

**Comment by msbrogli** [comment]
> @msbrogli I don't like it for the same reason we made the explorer 'Address Search' page so complex. An address might have many tokens and the balance is different for each other. The best option would be to have an address screen similar to the one from the explorer, but this is lot of work for now. I've changed the 'used' column for the 'number of transactions' column and added a link to the explorer, so the user can see all the details of the address. What do you think?

I agree, and I think your solution is good. Thanks!

**Comment by msbrogli** [comment]
Number 3 is the one I like the most, as the screenshot below:

image

File: src/App.js [hunk 1]
```diff
@@ -96,7 +97,9 @@ class Root extends React.Component {
     hathorLib.WebSocketHandler.removeListener('is_online', this.isOnlineUpdate);
     hathorLib.WebSocketHandler.removeListener('reload_data', this.reloadData);
 
-    IPC_RENDERER.removeAllListeners("ledger:closed");
+    if (IPC_RENDERER) {
```
Comment (msbrogli) [inline src/App.js:100]: Is it a bug fix? Maybe we should do it in another PR.

File: src/components/Navigation.js [hunk 2]
```diff
@@ -54,6 +54,9 @@ class Navigation extends React.Component {
               <li className="nav-item">
                 <NavLink to="/custom_tokens/" exact className="nav-link" activeClassName="active" activeStyle={{ fontWeight: 'bold' }}>{t`Custom tokens`}</NavLink>
               </li>
+              <li className="nav-item">
+                <NavLink to="/addresses/" exact className="nav-link" activeClassName="active" activeStyle={{ fontWeight: 'bold' }}>{t`Addresses`}</NavLink>
```
Comment (msbrogli) [inline src/components/Navigation.js:58]: If we go with number 3, we should remove this change.

File: src/screens/AddressList.js [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+  /**
+   * Return total number of transactions using an address
+   *
+   * @param {String} address Address to get the total number of transactions
+   * @param {Object} historyTransactions History of transactions from the wallet data storage
+   *
+   * @return {Number} Total number of transactions using this address
+   */
+  getNumberOfTransactions = (address, historyTransactions) => {
```
Comment (msbrogli) [inline src/screens/AddressList.js:75]: This algorithm seems really slow for big wallets because we are iterating all transactions multiple times, one for each address. I think it can receive the `addressKeys` and iterate only once storing the counters in a dictionary. Then, it returns the dictionary. Something like this:

```
counter = {};
for (const key in addressKeys) {
  counter[key] = 0;
}
```

Then, instead of `if (el.decoded.address === address)`, it just have to check if the key exists `if (el.decoded.address in counter)`.

**Comment by obiyankenobi** [comment]
I don't think option 3 is bad, but the problem I have with placing "See all addresses" there is that we're exposing something a bit more advanced, that people don't usually need to see. As we're debating about the gap addresses, it might lead to people having more doubts and feeling insecure/unsafe. That's why I think it should be placed on Settings page.

**Comment by pedroferreira1** [comment]
As discussed in the meeting today, we will go with number 3 and won't change anything for the GAP limit unused addresses.

File: src/App.js (hunk 1 above)
Comment (pedroferreira1) [reply src/App.js:100]: It's not exactly a bug fix. This raises an error only when running in the browser (so only for dev). I would prefer not to have a PR only for this but if you think it's important, I can do it.

File: src/components/Navigation.js (hunk 2 above)
Comment (pedroferreira1) [reply src/components/Navigation.js:58]: Removed

File: src/App.js (hunk 1 above)
Comment (msbrogli) [reply src/App.js:100]: I got it. No, I think it is good.

File: src/screens/AddressList.js (hunk 3 above)
Comment (pedroferreira1) [reply src/screens/AddressList.js]: Updated the code. Can you see if it's good?

File: src/screens/AddressList.js [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+   * @return {Number} Total number of pages of the list
+   */
+  getTotalPages = (array) => {
+    return Math.ceil(array.length / WALLET_HISTORY_COUNT);
+  }
+
+  /**
+   * Update number of transactions for each address
+   *
+   * @param {Object} addressData Transactions by address object {'address': 0}
```
Comment (msbrogli) [inline src/screens/AddressList.js:75]: I think `{'address': 0}` may be misleading. I'd use `{[address]: 0}`.

Comment (pedroferreira1) [reply src/screens/AddressList.js]: Changed
